pre_dispatch tracing: fix for nn.MultiheadAttention

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/106302, see the issue for details.

One minor risk in this PR is that: if you pre_dispatch trace `nn.MultiheadAttention`, **and** your inputs are some custom torch_function subclass, we'll incorrectly assume that the fast-path is okay to use. Given that pre_dispatch is only used underneath our `torch.export` and `torch.compile` subsystems, and it runs in a second pass **after** dynamo has captured the graph, then I think the risk here is pretty low - in both of those cases, dynamo should have already desugared away any torch_function subclasses/modes before we get to our second round of tracing to run with pre_dispatch.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106315
* #106461
* #106460

bdhirsh committed:
pre_dispatch tracing: fix for nn.MultiheadAttention

[ghstack-poisoned]

ezyang commented on torch/nn/modules/activation.py:
```diff
@@ -1162,7 +1162,11 @@ def forward(
-            if torch.overrides.has_torch_function(tensor_args):
+            torch_fn_mode_stack = torch.overrides._get_current_function_mode_stack()
+            is_pre_dispatch_tracing = len(torch_fn_mode_stack) == 1 and \
+                type(torch_fn_mode_stack[0]) == torch.fx.experimental.proxy_tensor.PreDispatchTorchFunctionMode
```
This part (where you test that there is only one function mode and it is this), should get factored into a helper function. This is too many lines here.

bdhirsh committed:
Update on "pre_dispatch tracing: fix for nn.MultiheadAttention"

Fixes https://github.com/pytorch/pytorch/issues/106302, see the issue for details.

One minor risk in this PR is that: if you pre_dispatch trace `nn.MultiheadAttention`, **and** your inputs are some custom torch_function subclass, we'll incorrectly assume that the fast-path is okay to use. Given that pre_dispatch is only used underneath our `torch.export` and `torch.compile` subsystems, and it runs in a second pass **after** dynamo has captured the graph, then I think the risk here is pretty low - in both of those cases, dynamo should have already desugared away any torch_function subclasses/modes before we get to our second round of tracing to run with pre_dispatch.




[ghstack-poisoned]

bdhirsh committed:
Update on "pre_dispatch tracing: fix for nn.MultiheadAttention"

Fixes https://github.com/pytorch/pytorch/issues/106302, see the issue for details.

One minor risk in this PR is that: if you pre_dispatch trace `nn.MultiheadAttention`, **and** your inputs are some custom torch_function subclass, we'll incorrectly assume that the fast-path is okay to use. Given that pre_dispatch is only used underneath our `torch.export` and `torch.compile` subsystems, and it runs in a second pass **after** dynamo has captured the graph, then I think the risk here is pretty low - in both of those cases, dynamo should have already desugared away any torch_function subclasses/modes before we get to our second round of tracing to run with pre_dispatch.




[ghstack-poisoned]

kimishpatel commented on test/test_proxy_tensor.py:
```diff
@@ -167,6 +167,27 @@ def forward(self, a_1):
+    _native_multi_head_attention = torch.ops.aten._native_multi_head_attention.default(arg2_1, arg2_1, arg2_1, 4, 4, _param_constant0, _param_constant1, _param_constant2, _param_constant3);  arg2_1 = _param_constant0 = _param_constant1 = _param_constant2 = _param_constant3 = None
+    getitem = _native_multi_head_attention[0]
+    getitem_1 = _native_multi_head_attention[1];  _native_multi_head_attention = None
+    return (getitem, getitem_1)""")  # noqa: B950
```
Dont you also want to add test when pre_dispatch=False, to make sure native_mult_head_attenttion actually gets decomposed?

guangy10 replied:
@bdhirsh From the model I'm working on, when `pre_dispatch=False`, the `_native_multi_head_attention` op still shows up in the graph and not gets decomposed.

kimishpatel commented:
> in both of those cases, dynamo should have already desugared away any torch_function subclasses/modes before we get to our second round of tracing to run with pre_dispatch

Why has dynamo already desugared torch_function subclasses? Is it because we already have run through torch_function subclasses when doing dynamo tracing, before make_fx?

bdhirsh commented:
> Why has dynamo already desugared torch_function subclasses? Is it because we already have run through torch_function subclasses when doing dynamo tracing, before make_fx?

Yep that's right (although I think support for handling all cases of __torch_function__ in dynamo isn't fully there yet, Michael Lazos was working on it though). Dynamo is fully responsible for tracing through __torch_function__, so by the time we get to pre_dispatch tracing in the export workflow, we shouldn't have to worry about any other __torch_function__ code from the user.

bdhirsh committed:
Update on "pre_dispatch tracing: fix for nn.MultiheadAttention"

Fixes https://github.com/pytorch/pytorch/issues/106302, see the issue for details.

One minor risk in this PR is that: if you pre_dispatch trace `nn.MultiheadAttention`, **and** your inputs are some custom torch_function subclass, we'll incorrectly assume that the fast-path is okay to use. Given that pre_dispatch is only used underneath our `torch.export` and `torch.compile` subsystems, and it runs in a second pass **after** dynamo has captured the graph, then I think the risk here is pretty low - in both of those cases, dynamo should have already desugared away any torch_function subclasses/modes before we get to our second round of tracing to run with pre_dispatch.




[ghstack-poisoned]

bdhirsh commented:
Closing this PR, since it isn't actually what the executorch team wants (they actually want to make sure that the fastpath op doesn't show up in their traces during export) Talked with @guangy10 on alternative, which is just to make sure that normal (non-pre_dispatch) tracing does not try to bake in the fast-path op.
